keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,945 @@
+Local policy, it was dropped later on. This meant that now
+TopologyAwareRouting in conjunction with InternalTrafficPolicy didn’t exactly
+allow users to express a much desired use case from topologyKeys which is
+"prefer node-local, failover to same zone, then route anywhere" While this
+specific behavior is a non-goal for the initial implementation of the
+`routingPreference` field, the design allows for the potential introduction of
+such a preference in future refinements.
+
+## Goals
+
+* **Enhanced User Control:** Provide users with a knob to specify a preference that
+  MAY influence the traffic routing of a Kubernetes Service.
+
+* **Guidance for Implementations:** Offer standard routing preference values
+  that implementations can choose to recognize and support, promoting a degree
+  of commonality.
+
+* **Flexibility:** Allow implementations to interpret and support standard
+  heuristics in a manner that aligns with their capabilities and design.
+
+* **Extensibility:** Enable Kubernetes implementations to introduce and support
+  innovative routing strategies (for example, those based on topology, latency,
+  or other custom heuristics) that can evolve alongside users needs and
+  infrastructure capabilities.
+
+## Non-Goals
+
+* **Strict Routing Guarantees:** The `routingPreference` field will not
+  enforce deterministic routing paths. It serves as a mechanism for expressing
+  hints and preferences that implementations can consider when making routing
+  decisions.
+
+* **Mandatory and Uniform Implementation Support:** Kubernetes implementations
+  are not required to support all standard heuristics (e.g., `Zone`, `Spread`).
+  Even when standard heuristics are supported, their precise behavior and
+  interpretation might vary across implementations.
+
+* **Replacement of Traffic Policies:** The new field is complementary to
+  InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute
+  their role in enforcing strict traffic locality.
+
+* **Immediate Support for All Possible Heuristics:** The initial implementation
+  focuses on a core set of heuristics. Addition of new heuristics (like
+  `Local` for Node local preference) could be explored in future
+  refinements.
+
+## Proposal
+
+Add a new field, `routingPreference`, to the Service specification. This field
+will serve as preference or hint for the underlying implementation to consider
+while making routing decisions. It does not offer strict routing guarantees.
+
+The field will support the following initial values:
+
+* `Default`: Indicates no specific routing preference. The user delegates the
+  routing decision to the implementation, allowing it to apply its best-effort
+  strategy.
+* `Spread`: Encourages an equal distribution of traffic across
+  endpoints, potentially spanning multiple zones (or regions).
+* `Zone`: Encourages routing traffic to endpoints within the same zone as
```

Comment:
I hate rehashing this, but I am still unsure.

I don't think we want to be too precise. Calling this "Zone" implies that "Region" and "Node" don't factor.  I don't think we want very many of these.  If "Zone" means anything significantly different than "use zone locality", the name is wrong or the implementation is wrong.

> Now, If we were to choose the heuristic name of Close right now (which at the moment would mean "route traffic within the same zone") and then later on we want to introduce an equivalent to Local (still without any feedback mechanism available to us), I don't think we would be able to say that, henceforth, Close means "route traffic within the same node" because that is increasing the risk. 

I don't think I agree.  If the new definition of "Close" means significantly more risk, the implementation should expose it with a new name.  But is that on US to define?  Adding a standard mode should have a VERY VERY high bar

Reply:
I don't think anyone wants a mode that prefers-but-does-not-require same-zone, but which does not also prefer-but-not-require same-region and same-node.

It seems to me like there are three kinds of users/services:
1. "I don't have any special requirements, please just route my services efficiently, and if you can change the routing rules to make it even more efficient in the next release then yay"
2. "I don't have any special requirements, but I don't want to wait for you to make service routing more efficient (or I think I'm smarter than you), so please give me back my `topologyKeys`."
3. "I have specific requirements, which may involve money, and want to be able to enforce those requirements precisely, without worrying that a future version of kube-proxy will optimize around them"

If we aren't going to humor group 2, then groups 1 and 2 collapse together. They basically want "send to the endpoint that best balances (a) low latency, (b) high throughput, (c) the future availability of low-latency high-throughput connections for all clients". Though they can benefit from vaguely-defined routing modes, which will help them express hints like "prioritize latency" (`Close`) or "prioritize throughput/reliability" (`Spread`) so "dumb" proxy implementations can get slightly better behavior, while "smarter" proxy implementations can balance those vague hints against what their own endpoint monitoring is telling them.

But group 3 primarily wants to avoid being charged for cross-zone traffic. If their cluster is in a situation where zone 1's endpoints are getting overloaded and zone 2's endpoints are idle, they don't want Kubernetes to automatically start routing traffic to zone 2 for them, they want Kubernetes to automatically deploy more endpoints to zone 1 for them. So they want modes that are extremely precise about some things, though they can still be vague about other things. (eg, if they say `RequireSameZone`, the proxy is still allowed to try to optimize endpoint balance _after_ it has enforced that constraint).

If we don't care about group 3, then I don't think there's any reason for the API to mention zones at all. It doesn't make sense to have a mode that means "keep traffic within the same zone if it makes sense, but let it go to other zones if that would be better", because that should just be part of `Default`.

Reply:
We think almost everyone should choose Default, right?  BUt we're positoning the others as "if I have to make a choice between good and fast, I want..."  

The user-story around helm charts is what keeps coming back to my mind.  If I know that I have set up zonal scaling and so forth, it's OK for me to choose "fast" over "good" (because I handled Goodness myself).

So maybe the question is whether this is really a PREFERENCE that implementations may ignore (I know you said prefer zone, but I know from feedback that this is bad) or strict?  If you really really want strict same-zone, maybe you should have a different Service per zone?  That's crappy for LBs, I guess.

So I'll ask again: does "Zone" mean an implementation must ignore information and make "wrong" choices?  Does "Spread" mean it must cross zones sometimes, just to say it approached a random distribution?

Reply:
> We think almost everyone should choose Default, right?

In the future, yes, almost everyone should choose Default (while they zip around Mars with their jetpacks). In the short run though, we know nobody is actually doing endpoint monitoring, and we don't have any specific plans to add it to kube-proxy, so hints seem useful. (But we want to define the hints so that future jetpack kubernetes isn't overly constrained by them.)

>  BUt we're positoning the others as "if I have to make a choice between good and fast, I want..."

zone is sometimes about "cheap", not just "fast"

> So maybe the question is whether this is really a PREFERENCE that implementations may ignore

There are people who at least claim that they want strict.

>  ("I know you said prefer zone, but I know from feedback that this is bad")

It was bad at the point you gathered the feedback, but what if their helm chart includes a HorizontalPodAutoscaler that is already in the process of fixing that?

We can assume that the service proxy is more clever than the average user, but we can't assume it's more clever than _every_ user...

---

Another (possibly implausible) "strict" example: "clients in the EU region must be routed to endpoints in the EU region, for GDPR compliance". If you actually had a cluster where a constraint like that was needed, then it's no longer even just about "more efficient" vs "less efficient"; you'd actually rather have connections get _dropped_ than be misrouted. (Which is probably not actually what most people want in the "RequireZone" case...)

Reply:
> If you actually had a cluster where a constraint like that was needed, then it's no longer even just about "more efficient" vs "less efficient"; you'd actually rather have connections fail than be misrouted. 

This is a good point and something that the current generation of hints are sadly unprepared to handle since we simply fall back to cluster-wide routing if no endpoints have matching zone hints. I think historically we've said that "RequireZone" and "RequireRegion" were sufficiently niche to be out of scope for this, but it's worth noting that not just are they out of scope, they'd require different implementation details (although of course it could just be new values added to this API in the future).

The very nature of the proposed field name here indicates that it's a preference and not a policy, similar conceptually to "hint" vs "requirement". I think that we might be able to make a case for something like "RequireZone" belonging as an option in internal and external traffic policy fields in the future instead of this preference field, but maybe that's an unnecessary and unhelpful distinction. In that world you'd have the following logic for each type of traffic:

_Is the policy field set to a non-"cluster" value for this type of traffic?_

Yes: Policy has precedence, policy is likely ignored unless it's somehow meaningful within the subset of endpoints the policy results in
No: Use the routing preference specified by this new field

Does that logic make sense? Is it weird to have external vs internal distinctions for traffic policies but not this routing preference?

For reference, what I've described above is pretty similar to [how we handle this today in kube-proxy](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/pkg/proxy/topology.go#L40) with hints, with the difference being that this leaves open the possibility that both a policy and a preference could both be applied for the same service which is not currently possible.